edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,131 @@
+use futures::executor::block_on;
+use futures::join;
+use nix::sys::signal::{self, Signal};
+use nix::unistd::Pid;
+use signal_hook::{iterator::Signals, SIGINT, SIGTERM};
+use std::io::Error;
+use std::process::{exit, Child, Command, Stdio};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+fn main() -> Result<(), Error> {
+    block_on(async_main())
+}
+
+async fn async_main() -> Result<(), Error> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = match register_sigterm_listener() {
+        Ok(should_shutdown) => should_shutdown,
+        Err(e) => {
+            error!("Failed to register sigterm listener. Shutting down.");
+            return Err(e);
+        }
+    };
+
+    // start edgehub and blow up if can't start
+    let mut edgehub = Command::new("dotnet")
+        .arg("/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll")
+        .stdout(Stdio::inherit())
+        .spawn()
+        .expect("Failed to start Edge Hub process");
+    info!("Launched Edge Hub process with pid {:?}", edgehub.id());
+
+    // unwrap broker if started, else shutdown edgehub and exit
+    let maybe_broker = match Command::new("/usr/local/bin/mqttd")
+        .stdout(Stdio::inherit())
+        .spawn()
+    {
+        Ok(child) => Some(child),
+        Err(e) => {
+            error!("Failed to start MQTT Broker process. {:?}", e);
+            None
+        }
+    };
+    let mut broker = maybe_broker.unwrap_or_else(|| {
+        info!("Broker process not started, so shutting down EdgeHub and exiting");
+        let shutdown = shutdown(&mut edgehub);
+        block_on(shutdown);
+        exit(1);
```

Comment:
assuming exit code 1 is fine

Reply:
let's make a different exit code when edgehub and broker and watchdog exits with error?